lib/ansible/modules/cloud/vmware/vmware_guest_serial_port.py

```diff
@@ -0,0 +1,439 @@
+    - type: 'file'
+      state: 'new'
+      file_path: 'newfile'
+      yield_on_poll:  True
+    force: True
+    register: create_multiple_ports
+
+# Modify existing serial port
+- name: Modify Network backing type
+  vmware_serial_port:
+    hostname: '{{ vcenter_hostname }}'
+    username: '{{ vcenter_username }}'
+    password: '{{ vcenter_password }}'
+    name: '{{ name }}'
+    backings:
+    - type: 'network'
+      state: 'present'
+      direction: 'server'
+      service_uri: 'tcp://1000'
+  delegate_to: localhost
+
+# Remove serial port
+- name: Remove pipe backing type
+  vmware_serial_port:
+    hostname: '{{ vcenter_hostname }}'
+    username: '{{ vcenter_username }}'
+    password: '{{ vcenter_password }}'
+    name: '{{ name }}'
+    backings:
+    - type: 'pipe'
+      state: 'absent'
+  delegate_to: localhost
+
+'''
+
+RETURN = r'''
+serial_port_data:
+    description: metadata about the virtual machine's serial ports after managing them
+    returned: always
+    type: dict
+    sample: [
+        {
+          "backing_type": "network",
+          "direction": "client",
+          "service_uri": "tcp://6000"
+        },
+        {
+          "backing_type": "pipe",
+          "direction": "server",
+          "pipe_name": "serial pipe"
+        },
+    ]
+'''
+
+from ansible.module_utils.basic import AnsibleModule
+from ansible.module_utils.vmware import PyVmomi, vmware_argument_spec, set_vm_power_state, wait_for_task, gather_vm_facts
+from ansible.module_utils._text import to_native
+try:
+    from pyVmomi import vim
+except ImportError:
+    pass
+
+
+class PyVmomiHelper(PyVmomi):
+    """ This class is a helper to create easily VMware Spec for PyVmomiHelper """
+
+    def __init__(self, module):
+        super(PyVmomiHelper, self).__init__(module)
+        self.change_applied = False   # a change was applied meaning at least one task succeeded
+        self.config_spec = vim.vm.ConfigSpec()
+        self.config_spec.deviceChange = []
+
+    def check_vm_state(self, vm_obj):
+        """
+        To add serial port, the VM must be in powered off state
+
+        Input:
+          - vm: Virtual Machine
+
+        Output:
+          - [proceed, current_state]
+        """
+        facts = gather_vm_facts(self.content, vm_obj)
+        current_state = facts['hw_power_status'].lower()
+        if current_state == "poweredoff":
+            return [True, current_state]
+        elif self.params.get('force'):
+            set_vm_power_state(self.content, vm_obj, 'poweredoff', True)
+            return [True, current_state]
+        else:
+            return [False, current_state]
+
+    def get_serial_port_config_spec(self, vm_obj):
+        """
+        Variables changed:
+          - self.config_spec
+          - self.change_applied
+        """
+        # create serial config spec for adding, editing, removing
+        for backing in self.params.get('backings'):
+            if backing['state'].lower() == 'new':
```

Comment:
I don't like the `new` state because it breaks the idempotency of the module. It sounds like you can use `service_uri`, `device_name` and `file_path` to know if a device is already present. Could you do that instead?

Reply:
@goneri 
How about the usecase when the user wants to attach 2 serial devices of same backing type?

Reply:
I believe it should be type+device_name/service_uri/file_path.